rrtmgp/mo_gas_concentrations.F90

```diff
@@ -142,18 +143,27 @@ function set_vmr_scalar(this, gas, w) result(error_msg)
     if (associated(this%concs(igas)%conc)) then
       if ( any(shape(this%concs(igas)%conc) /= [1, 1]) ) then
         !$acc exit data delete(this%concs(igas)%conc)
+        !$omp target exit data map(release:this%concs(igas)%conc)
         deallocate(this%concs(igas)%conc)
         nullify   (this%concs(igas)%conc)
       end if
     end if
     if (.not. associated(this%concs(igas)%conc)) then
       allocate(this%concs(igas)%conc(1,1))
       !$acc enter data create(this%concs(igas)%conc)
+      !$omp target enter data map(alloc:this%concs(igas)%conc)
     end if
 
+    p => this%concs(igas)%conc(:,:)
     !$acc kernels
+    !$omp target map(to:w)
+#ifdef _CRAYFTN
```

Comment:
Is the coding especially tricky here, or is this a weakness/bug in the Cray compiler?

Reply:
Yes. This is the most problematic file to create an OpenMP GPU version. 

The PGI OpenACC compiler is miles ahead of the Cray OpenMP compiler here. Suppose you have a derived type `foo` which contains another derived type `bar` and several allocatable components. It simply cannot handle mapping `foo%bar%comp` to the GPU. It doesn't really know how to handle that. I think the OpenMP compilers will get better at this in the future though.